src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -37,15 +38,53 @@ pub struct ObligationForest<O> {
     /// at a higher index than its parent. This is needed by the
     /// backtrace iterator (which uses `split_at`).
     nodes: Vec<Node<O>>,
-    snapshots: Vec<usize>
+    snapshots: Vec<usize>,
+
+    /// List of inversion actions that may be performed to undo mutations
+    /// while in snapshots.
+    undo_log: Vec<UndoAction<O>>,
```

Comment:
The pattern I used elsewhere is to have one vector and to push `OpenSnapshot` events into it. This way you only need one vector (the vector is empty if no snapshots are active). Committing the base snapshot clears the vector (but leaves the state the same). Converting a non-base snapshot converts the `OpenSnapshot` to `CommittedSnapshot`. Reverting a snapshot pops things off of the undo vector. If you ever pop off an `OpenSnapshot` (as opposed to a `CommittedSnapshot`)...then something went wrong. You can see an example in `src/librustc/middle/infer/region_inference.rs`.